File: src/model/Collections/MyDictionary.java
package model.Collections;

import interfaces.IDictionary;

import java.lang.reflect.Array;
import java.util.Arrays;

/**
 * Created by Lucian on 10/12/2015.
 */
public class MyDictionary<K, V> implements IDictionary<K, V> {

    private Tuple[] array;
    private int size;

    private class Tuple{

        private K key;
        private V value;

        public Tuple(K key, V value) {
            this.key = key;
            this.value = value;
        }


        public K getKey() {
            return key;
        }

        public void setKey(K key) {
            this.key = key;
        }

        public V getValue() {
            return value;
        }

        public void setValue(V value) {
            this.value = value;
        }
    }

    public MyDictionary(int size) {
        this.size = 0;
        Object[] tmp = new Object[size];
        array = Arrays.copyOf(tmp, tmp.length, Tuple[].class);
    }

    @Override
    public V lookUp(K id) {
        for(int i =0; i < size; i++){
            if (array[i].getKey().equals(id)){
                return array[i].getValue();
            }
        }
        return null;
    }

    @Override
    public boolean isEmpty() {
        return size <= 0;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean containsKey(K key) {
        for(int i =0; i < size; i++){
            if (array[i].getKey().equals(key)){
                return true;
            }
        }
        return false;
    }

    @Override
    public void put(K key, V value) {
        boolean found = false;
        for(int i =0; i < size; i++){
            if (array[i].getKey().equals(key)){
                array[i].setValue(value);
                found = true;
            }
        }

        if(!found){
            array[size] = new Tuple(key, value);
            size++;
        }
    }

    @Override
    public void remove(K key) {
        for(int i =0; i< size; i++){
            if(array[i].getKey().equals(key)){
                for(int j = i; j < size -1; j++){
                    array[j] = array[j+1];
                }
                size --;
                break;
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for(int i =0; i< size; i++){
            result.append(array[i].key + "->" + array[i].value+"\n");
        }
        return result.toString();
    }
}

File: C#/ToyLanguage/ToyLanguage/Model/Collections/MyDictionary.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguage.Interfaces;

namespace ToyLanguage.Model.Collections
{
    [Serializable]
    class MyDictionary<K, V> : IMyDictionary<K, V>
    {
        /**
    * An array of tuple (key, value)
    */
        private Tuple[] array;

        /**
         * The size of the dictionary
         */
        private int sizeOfArray;

        /**
         * Inner class Tuple
         */
        private class Tuple
        {

            /**
             * The key
             */
            public K key;

            /**
             * The value
             */
            public V value;

            /**
             * The constructor
             *
             * @param key   The key
             * @param value The value
             */
            public Tuple(K key, V value)
            {
                this.key = key;
                this.value = value;
            }

            /**
             * Get the key
             *
             * @return Key
             */
            public K getKey()
            {
                return key;
            }

            /**
             * Set the key
             *
             * @param key Key
             */
            public void setKey(K key)
            {
                this.key = key;
            }

            /**
             * Get the value of current key
             *
             * @return Value
             */
            public V getValue()
            {
                return value;
            }

            /**
             * Set the value of the current key
             *
             * @param value Value
             */
            public void setValue(V value)
            {
                this.value = value;
            }
        }

        /**
         * Constructor
         *
         * @param size int max size of dictionary
         */
        public MyDictionary(int cap)
        {
            sizeOfArray = 0;
            Object[] tmp = new Object[cap];
            array = new Tuple[cap];
           
         }

    /**
     * Get the value of the id
     *
     * @param id The id
     * @return Value
     */
    public V lookUp(K id)
    {
        for (int i = 0; i < sizeOfArray; i++)
        {
            if (array[i].getKey().Equals(id))
            {
                return array[i].getValue();
            }
        }

       //if the key is not here return null
        throw new System.ArgumentException("Key not in table", "original");
    }

    /**
     * Check if dictionary is empty
     *
     * @return True if yes, false otherwise
     */
    public bool isEmpty()
    {
        return sizeOfArray <= 0;
    }

    /**
     * Get the size of the dictionary
     *
     * @return int size
     */
    public int size()
    {
        return sizeOfArray;
    }

    /**
     * Check if K key is in dictionary
     *
     * @param key The key
     * @return True if yes, false otherwise
     */

    public bool containsKey(K key)
    {
        for (int i = 0; i < sizeOfArray; i++)
        {
            if (array[i].getKey().Equals(key))
            {
                return true;
            }
        }
        return false;
    }

    /**
     * Put the V value in K key
     *
     * @param key   The key
     * @param value The value
     */
 
    public void put(K key, V value)
    {
        bool found = false;
        for (int i = 0; i < sizeOfArray; i++)
        {
            if (array[i].getKey().Equals(key))
            {
                array[i].setValue(value);
                found = true;
            }
        }

        if (!found)
        {
            array[sizeOfArray] = new Tuple(key, value);
            sizeOfArray++;
        }
    }

    /**
     * Remove the key from dictionary
     *
     * @param key The key
     */
    public void remove(K key)
    {
        for (int i = 0; i < sizeOfArray; i++)
        {
            if (array[i].getKey().Equals(key))
            {
                for (int j = i; j < sizeOfArray - 1; j++)
                {
                    array[j] = array[j + 1];
                }
                sizeOfArray--;
                break;
            }
        }
    }

    /**
     * String representation
     *
     * @return String
     *
     */
    public string ToString()
    {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < sizeOfArray; i++)
        {
            result.Append(array[i].key + "->" + array[i].value + "\n");
        }
        return result.ToString();
    }

        public void addAll(Dictionary<K, V> all)
        {
            throw new NotImplementedException();
        }

        public Dictionary<K, V> getMap()
        {
            throw new NotImplementedException();
        }
    }
}
